NIFI-2078, 2363, 2364, 2442: External state management. CLUSTER scope behavior in cluster. Clear state error handling.

ijokarumawak opened the pull request:
- Added ExternalStateManager to handle components' state managed
  externally
- Added UI codes to display external state
- Added view/clear functionality to ConsumeKafka
- Added view/clear functionality to GetKafka
- Capture property value change, so that external state can be accessed
  before onTrigger is called
- Fixed Component State UI clear link NIFI-2078
- Added Expression Language support at onPropertyModified.
- Tested with Kerberized Zookeeper and Kafka

ijokarumawak commented:
Integration test with Kerberized Kafka and Zookeeper haven't finished yet, but the implementation is done. Would you please initiate the code review process?

This PR added External State management functionality to GetKafka and ConsumeKafka as following screenshot:

![componentstate-consumekafka](https://cloud.githubusercontent.com/assets/1107620/16295638/f31a3c1e-3962-11e6-9b71-63f4f581e199.png)

ijokarumawak commented:
Test with Kerberized Kafka and Zk is done. I followed [confluent Apache Kafka Security 101](http://www.confluent.io/blog/apache-kafka-security-authorization-authentication-encryption) to setup test environment. As written in the blog post, I used the [Vagrant instance](https://github.com/confluentinc/securing-kafka-blog), then deployed NiFi Snapshot build on it.
In order to make ConsumeKafka working, followings had to be configured:

- $NIFI/conf/bootstrap.conf
  - java.arg.15=-Djava.security.auth.login.config=/etc/kafka/kafka_client_jaas.conf
- ConsumeKafka processor properties
  - Kafka Brokers: kafka.example.com:9095
  - Topic Name: securing-kafka
  - Client ID: (any value would be fine)
  - Security Protocol: SASL_SSL
  - Kerberos Service Name: kafka
  - Group ID: securing-kafka-group * needs ACL setting
  - Offset Reset: earliest (doesn't matter)
  - Message Demarcrator: (No value set)
  - ssl.key.password: test1234
  - ssl.keystore.location: /etc/security/tls/kafka.client.keystore.jks
  - ssl.keystore.password: test1234
  - ssl.truststore.location: /etc/security/tls/kafka.client.keystore.jks
  - ssl.truststore.password: test1234

Thanks @olegz for helping me to setup the environment. I also should thank [confluent](http://www.confluent.io/) for publishing the detailed blog post and share Vagrant file!

JPercivall commented:
Starting to review.

Unless if I'm missing something, why is external state only added for GetKafka and ConsumeKafka? Aren't there many other processors that manage state externally?

ijokarumawak commented:
@JPercivall Thanks for your time to review.

I've looked at every existing Processors, Reporting tasks, and Controller services, but couldn't find other component needing external state management from NiFi.

If you know particular component should implement this, please let me know.

JPercivall commented:
@ijokarumawak I believe you're right about it only being Kafka. I looked through them as well and couldn't find an instance of external state. 

While surprised I am definitely happy because it makes this PR a lot easier to review, lol.

JPercivall commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/main/java/org/apache/nifi/processors/kafka/GetKafka.java:
```diff
@@ -184,6 +195,9 @@
     private volatile long deadlockTimeout;
 
     private volatile ExecutorService executor;
+    private String zookeeperConnectionString;
```
These three will be accessed by different threads and should be marked volatile.

ijokarumawak replied:
Addressed.

JPercivall commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/main/java/org/apache/nifi/processors/kafka/GetKafka.java:
```diff
@@ -218,11 +232,13 @@
 
     public void createConsumers(final ProcessContext context) {
-        final String topic = context.getProperty(TOPIC).getValue();
+        topic = context.getProperty(TOPIC).getValue();
```
Why change the three values here? They are set in the "onPropertyModified" so whenever it's changed it will readjust.

ijokarumawak replied:
I was just not brave enough to remove those codes. Confirmed it works without setting here. Thanks for pointing this out.

JPercivall commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/main/java/org/apache/nifi/processors/kafka/GetKafka.java:
```diff
@@ -481,4 +496,51 @@ private void releaseFlowFile(FlowFile flowFile, ProcessSession session, Map<Stri
+        if (!isReadyToAccessState()) {
+            return;
+        }
+        KafkaUtils.clearPartitionOffsets(zookeeperConnectionString, topic, groupId);
```
Shouldn't this block the onTrigger from attempting to reach out to Kafka to get new messages?

ijokarumawak replied:
Thanks, good point, I've added synchronized block in clearExternalState().

olegz commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/main/java/org/apache/nifi/processors/kafka/GetKafka.java:
```diff
@@ -481,4 +496,51 @@ private void releaseFlowFile(FlowFile flowFile, ProcessSession session, Map<Stri
+                || StringUtils.isEmpty(groupId)) {
+            return false;
+        }
+        return true;
```
Wouldn't it be simpler to do:
```java
return !StringUtils.isEmpty(zookeeperConnectionString) && !StringUtils.isEmpty(topic) && StringUtils.isEmpty(groupId);
```

ijokarumawak replied:
Yes, it would. Addressed, thanks!

JPercivall commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/test/java/org/apache/nifi/processors/kafka/TestGetKafka.java:
```diff
@@ -166,4 +170,44 @@ public MessageAndMetadata answer(InvocationOnMock invocation) throws Throwable {
+        runner.setProperty(GetKafka.GROUP_ID, "consumer-group-id");
+
+        try {
+            processor.getState();
```
This is a unit test to clearState but calls getState here.

ijokarumawak replied:
Totally a mistake, thanks! Addressed.

JPercivall commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-pubsub-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/ConsumeKafka.java:
```diff
@@ -239,4 +300,188 @@ private void releaseFlowFile(FlowFile flowFile, ProcessContext context, ProcessS
+        }
+
+        final String groupId = kafkaProperties.getProperty(ConsumerConfig.GROUP_ID_CONFIG);
+        final Boolean result = submitConsumerGroupCommand("Clear offsets", consumer -> {
```
Same comment as on GetKafka, should this block the onTrigger from getting new messages?

ijokarumawak replied:
Added synchronized block here as well.

JPercivall commented on nifi-api/src/main/java/org/apache/nifi/components/state/ExternalStateManager.java:
```diff
@@ -0,0 +1,63 @@
+     *
+     * @throws IOException if unable to communicate with the underlying storage mechanism
+     */
+    void clear() throws IOException;
```
This should be renamed to "clearState" to match the corresponding "getState" and to better identified the method. Since it's an interface you never know what else the processor may want to clear and "clear" could end up being ambiguous.

JPercivall commented:
This currently has conflicts with master and needs to be rebased before it can be merged (sorry for taking so long to review the PR).

ijokarumawak commented:
@JPercivall @olegz Thanks for reviewing this! I've rebased the branch, and addressed the feedback. However, while I was testing this again, I noticed that if it's running on a clustered NiFi environment, every node gets and clears offsets from Kafka. While it's working, I think these method should be executed on only the primary node, or if it's a standalone instance since offsets information is a global state among consumers.

With `@OnPrimaryNodeStateChange` annotation, it's possible to know if it's a cluster and a primary node. But it's only called if a node get elected as a primary, or a node was primary but revoked. Since other node is not notified, I couldn't distinguish standalone or primary from a processor..
So, I left it as it is for now. Do you know If there's a way to know if NiFi is clustered from a processor?

olegz commented:
@ijokarumawak I am not yet done, but good that you have rebased. I am just tied up with another thing at the moment, but will be reviewing later on

ijokarumawak commented:
@JPercivall @olegz @markap14 @mcgilman 

I'd like to hear your thoughts.
I've been researching how state management works in a cluster environment and found out:

1. LOCAL scope `Set` is executed on every node from onTrigger
2. LOCAL scope `View state` result will be merged at `ComponentStateEndpointMerger`, so each node can return its own state in a single merged response
3. LOCAL scope `Clear state` is executed on every node, then one of the successful results will be returned. If it has a mixed result of success and failure, current implementation returns success result.
4. CLUSTER scope `Set` is usually done from onTrigger, such as ListXXX processors, and we recommend to use `on primary node` schedule strategy. So usually it's not a problem. But both DFM and developer be careful.
5. CLUSTER scope `View state` is executed on every node, even if the same state will be returned from Zookeeper. Then `ComponentStateEndpointMerger` uses one of those results. I think this should be executed **only on the primary node**.
6. CLUSTER scope `Clear state` is executed on every node. Updating the same resource in Zk. I think this should be executed **only on the primary node**.
7. EXTERNAL scope `Set` happens externally, NiFi doesn't control it.
8. EXTERNAL scope `View state` is currently executed on every node, but for Kafka, I think it should be executed on **only on the primary node**. However, there may be components which retrieves node dependent EXTERNAL state in the future. In that case, differentiator something like `@OnPrimaryNodeOnly` annotation can be added, but not required for now.
9. EXTERNAL scope `Clear state`: same as `View state`. Should be **only on the primary node**.

While it's working, I'm not comfortable with accessing external system like Zk or Kafka if it's unnecessary, also clearing same information from multiple nodes sounds problematic. I'd like to modify CLUSTER scope behavior as well, in addition to EXTERNAL scope, to limit those happen on primary node only, at StandardNiFiServiceFacade and DAO layers.

How do you think?

markap14 commented:
@ijokarumawak I am not opposed to performing the view / clear on primary node only when NiFi is in clustered mode. That wouldn't have been a good idea in the 0.x baseline because the primary node may be disconnected. But in 1.0 we would auto-elect a new primary so I am ok with it in the 1.x baseline. For external state, though, we would need some sort of indicator as to whether the state should be cleared on all nodes or only the primary node. This could perhaps be part of the ExternalStateManager interface - a "Scope getStateScope()" method so that we know if external state is being managed per-node or cluster-wide. This would tell us if we need to replicate the request to all nodes or only the primary node.

ijokarumawak commented:
Thanks @markap14 . I understand, I'll test to see if this works well with primary node failure and auto-elect scenario. It'd be the best if we could control nodes we replicate the HTTP requests to, however, getState and clearState are designed as per component request, meaning, local, cluster and external state are accessed all at once. So I will simply replicate HTTP requests to all nodes as it does now, and check the external state scope at DAO or service layer.

mcgilman commented:
@ijokarumawak It appears the 'Clear state' link has been moved in an unrelated effort to above the table on the right hand side. It looks like your PR doesn't adjust the positioning or z-index of the link so I'm going to assume this PR doesn't make changes to the link and that NIFI-2078 is already resolved. Just looking to resolve and wrap up that JIRA. Thanks!

ijokarumawak commented:
@mcgilman Yes, I noticed that 'Clear state' moved above the state table, so I removed z-index change. Thanks :)

mcgilman commented:
@ijokarumawak Awesome, thank you. There was just some confusion about whether we could resolve the PR for NIFI-2078 and just wanted to verify with you first.

ijokarumawak commented:
Hello @JPercivall @olegz @markap14 I've updated Cluster and External state behavior in a NiFi cluster, so that only a Primary node access Zookeeper or External data store. 

Tested, with the latest master branch. Would you be able to resume review process for this?

Now this PR contains three related JIRA issues:
- [NIFI-2078](https://issues.apache.org/jira/browse/NIFI-2078): State managed externally
- [NIFI-2363](https://issues.apache.org/jira/browse/NIFI-2363): CLUSTER scope should be managed on primary node
- [NIFI-2364](https://issues.apache.org/jira/browse/NIFI-2364): Avoid being disconnected by error related to external system

Code change summaries for each topic is written in the commit message.

For External state, it's also possible to make the state external local (per node). Please see following screenshots, fig-1 and 2.

fig-1: External - CLUSTER
![External - CLUSTER](https://cloud.githubusercontent.com/assets/1107620/17093451/5be50578-5284-11e6-8c74-f370a75c43da.png)

fig-2: External - LOCAL, GetKafka in this image is just an example. The actual code uses External - CLUSTER scope. There's no implementation using External - LOCAL as of now
![External - LOCAL](https://cloud.githubusercontent.com/assets/1107620/17093484/7f60e3dc-5284-11e6-87cc-b0acc0c4a861.png)

Thanks!

JPercivall commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/http/endpoints/ClearComponentStateEndpointMerger.java:
```diff
@@ -0,0 +1,71 @@
+public class ClearComponentStateEndpointMerger extends AbstractSingleEntityEndpoint<ClearComponentStateResultEntity> {
+    public static final Pattern PROCESSOR_STATE_URI_PATTERN = Pattern.compile("/nifi-api/processors/[a-f0-9\\-]{36}/state/clear-requests");
+    public static final Pattern CONTROLLER_SERVICE_STATE_URI_PATTERN = Pattern.compile("/nifi-api/controller-services/[a-f0-9\\-]{36}/state/clear-requests");
+    public static final Pattern REPORTING_TASK_STATE_URI_PATTERN = Pattern.compile("/nifi-api/reporting-tasks/[a-f0-9\\-]{36}/state/clear-requests");
```
Hard coding these URI patterns feels wrong. Any changes to these URIs in another PR could mess this up entirely without anyone noticing.

JPercivall replied:
Ah I see that's how it's done in all the classes in the same package

ijokarumawak replied:
Yes, that's how it's done today. There may be some way to get URI patterns align with corresponding Resource class methods, though.

JPercivall commented on nifi-api/src/main/java/org/apache/nifi/components/state/ExternalStateManager.java:
```diff
@@ -0,0 +1,90 @@
+ * <p>
+ * This mechanism is designed to allow developers to easily store and retrieve small amounts of state.
+ * Since implementation of this interface interacts with remote system, one should consider the cost of
+ * retrieving this data, and the amount of data should be kept to the minimum required.
```
This paragraph doesn't make sense for External state since the user probably isn't choosing what to store

ijokarumawak replied:
Updated, please check the latest commit.

JPercivall commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/ClusterProtocolHeartbeatMonitor.java:
```diff
@@ -91,7 +92,7 @@ public ClusterProtocolHeartbeatMonitor(final ClusterCoordinator clusterCoordinat
 
         String hostname = properties.getProperty(NiFiProperties.CLUSTER_NODE_ADDRESS);
-        if (hostname == null) {
+        if (StringUtils.isEmpty(hostname)) {
```
This change will conflict with @markap14 change in this PR: https://github.com/apache/nifi/pull/688/files

ijokarumawak replied:
Thanks for pointing this out. I will remove these changes from this PR. Posted my comments on #688.

JPercivall commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-pubsub-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/ConsumeKafka.java:
```diff
@@ -54,7 +72,12 @@
+        " Offsets can be cleared when there is no consumer subscribing with the same consumer group id." +
+        " It may take more than 30 seconds for a consumer group to become able to be cleared after it is stopped from NiFi." +
+        " Once offsets are cleared, ConsumeKafka will resume consuming messages based on Offset Reset configuration.")
+public class ConsumeKafka extends AbstractKafkaProcessor<Consumer<byte[], byte[]>> implements ExternalStateManager {
```
Since "ExternalStateManager" is something that the developer will need to implement if they will use it, this should be documented in the Developer guide

ijokarumawak replied:
Added documentation in Developer Guide.

JPercivall commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-pubsub-processors/pom.xml:
```diff
@@ -35,6 +35,10 @@
         </dependency>
         <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-expression-language</artifactId>
```
This module didn't already have access to EL?

ijokarumawak replied:
I added EL dependency to use StandardPropertyValue class at ConsumeKafka.onPropertyModified.

Since ConsumeKafka allows EL for properties that needed to get state from Kafka, EL has to be evaluated at onPropertyModified, too. But to do so, we need PropertyValue instance, which is not available at onPropertyModified because it only receives String representation of oldValue and newValue. An implementation class of PropertyValue is not included in nifi-api. So, I needed to add the EL dependency.

However, I felt it's a little bit strange to support EL for things like 'topic' or 'bootstrap_servers', since ConsumeKafka doesn't take input flow files, and these properties can not be changed after ConsumerKafka connects to Kafka. It keep using the same kafkaResource instance. So, those property values are more static, rather than dynamically evaluated. It makes sense to support EL for those values for PublishKafka, though.

ijokarumawak replied:
BOOTSTRAP_SERVERS, TOPIC, CLIENT_ID are the properties shared among Consume/ProduceKafka. If we can separate and let Consume, ProduceKafka have those as its own property, ProduceKafka to support EL but not ConsumeKafka, then the dependency can be removed.

But existing flow configuration may have to be updated if one uses EL for those properties of ConsumeKafka. 

@olegz @JPercivall How do you think?

JPercivall replied:
@ijokarumawak supporting EL for those properties makes sense. There are subject-less functions in EL as well as the ability to reference system, environment and custom variables set through the variable registry by using EL. 

I was just confused because since they already had references to EL functions, why did this need to get added as an explicit dependency?

ijokarumawak replied:
@JPercivall I agree, it does make sense to use EL subject-less functions.

EL functions can be used from Processor through interfaces provided by nifi-api. But I needed an implementation StandardPropertyValue class to evaluate EL from onAttributeModified method. I think we can not eval EL from String. Can we?

ijokarumawak replied:
We may be able to improve the framework to provide onPropertyModified a PropertyValue which is already setup. I think valuable registry can be useful with env configs like broker address. Raised an issue [NIFI-2364](https://issues.apache.org/jira/browse/NIFI-2364).

JPercivall commented:
@olegz, when you get a chance could you review the Get/Consume Kafka changes in this PR?

olegz commented:
@JPercivall +1 on Kafka stuff

ijokarumawak commented:
@JPercivall Added another commit to address feedback, added more documentations. Could you review the changes? Thanks!

JPercivall commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/main/java/org/apache/nifi/processors/kafka/GetKafka.java:
```diff
@@ -481,4 +493,59 @@ private void releaseFlowFile(FlowFile flowFile, ProcessSession session, Map<Stri
+        } else if (TOPIC.equals(descriptor)) {
+            topic = newValue;
+        } else if (GROUP_ID.equals(descriptor)) {
+            groupId = newValue;
```
I ran into an error due to how GroupId is getting set. This only gets called if someone modifies the groupid but it has a default value. So it will be null if the user never modifies it and they will hit this error:

2016-07-29 13:24:12,283 WARN [Timer-Driven Process Thread-6] o.apache.nifi.processors.kafka.GetKafka GetKafka[id=3798f2c6-0156-1000-0000-0000138083b7] Processor Administratively Yielded for 1 sec due to processing failure
2016-07-29 13:24:12,283 WARN [Timer-Driven Process Thread-6] o.a.n.c.t.ContinuallyRunProcessorTask Administratively Yielding GetKafka[id=3798f2c6-0156-1000-0000-0000138083b7] due to uncaught Exception: java.lang.IllegalStateException: java.util.concurrent.ExecutionException: java.lang.NullPointerException
2016-07-29 13:24:12,284 WARN [Timer-Driven Process Thread-6] o.a.n.c.t.ContinuallyRunProcessorTask 
java.lang.IllegalStateException: java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at org.apache.nifi.processors.kafka.GetKafka.onTrigger(GetKafka.java:367) ~[na:na]
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27) ~[nifi-api-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1060) ~[nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:136) [nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:47) [nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:123) [nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_74]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_74]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_74]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_74]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_74]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_74]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_74]
Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) [na:1.8.0_74]
	at java.util.concurrent.FutureTask.get(FutureTask.java:206) [na:1.8.0_74]
	at org.apache.nifi.processors.kafka.GetKafka.onTrigger(GetKafka.java:360) ~[na:na]
	... 12 common frames omitted
Caused by: java.lang.NullPointerException: null
	at java.util.Hashtable.put(Hashtable.java:459) ~[na:1.8.0_74]
	at java.util.Properties.setProperty(Properties.java:166) ~[na:1.8.0_74]
	at org.apache.nifi.processors.kafka.GetKafka.createConsumers(GetKafka.java:238) ~[na:na]
	at org.apache.nifi.processors.kafka.GetKafka$1.call(GetKafka.java:355) ~[na:na]
	at org.apache.nifi.processors.kafka.GetKafka$1.call(GetKafka.java:352) ~[na:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_74]
	... 3 common frames omitted

JPercivall replied:
Probably the easiest way to address this is to set groupId when the default gets created for the property.

ijokarumawak replied:
Thanks for pointing this out. Addressed.
Fixed bootstrap server address for ConsumeKafka, too.

JPercivall commented:
Hey @ijokarumawak, I tried testing this using Consume/Publish Kafka with Kafka 0.10.0 but I'm not seeing any state come up with I view state for ConsumeKafka. In the below screenshots you can see the data is following but no state is being stored. I have tried all 3 values of the "offset reset" for ConsumeKafka but that doesn't make a difference. Is there something special I should be doing?

![screen shot 2016-07-29 at 12 54 58 pm](https://cloud.githubusercontent.com/assets/11302527/17256496/c2e0005c-558b-11e6-8616-ec40ded55155.png)
![screen shot 2016-07-29 at 12 54 09 pm](https://cloud.githubusercontent.com/assets/11302527/17256501/cb3bd4ba-558b-11e6-896e-378f8a08c51f.png)

ijokarumawak commented:
@JPercivall I was able to reproduce the behavior with Kafka 0.10.0. But figured out the cause comes from a different change.

[NIFI-2355](https://issues.apache.org/jira/browse/NIFI-2355) and its PR #715 added permission check to table sort comparator, in nf-component-state.js. I think it shouldn't be added to state table, since a given user won't have record level permission difference for component state.

#715 has the same change for changes nf-controller-services.js and other JS files. I assume adding the permission change to nf-component-state.js was a mistake. By removing the change from nf-component-state.js, I was able to view external state with Kafka 0.10.0 Broker.

It's worth noting that I got the same JS error with ListFile when I viewed its state, so the problem is not limited to external state.

Failed NIFI-2441 for that issue.

ijokarumawak commented:
@JPercivall Rebased, and added default value handling. Confirmed that I can see state with Kafka 0.10.0 broker.  It seems that newer consumer library won't work with older broker, so you may need to upgrade Kafka broker (server process) to 0.10.0 as well. [1]

[1] https://groups.google.com/forum/#!topic/confluent-platform/jdv10HahRd8

ijokarumawak commented:
@JPercivall I've tested with broker 0.9.0.1 and 0.10.0. Both worked, consumed messages with ConsumeKafka and GetKafka, also view/clear state.

ijokarumawak commented:
@JPercivall I've changed how to capture property value changes. Added validation method in ExternalStateManager to validate and capture property values with a validation context.

It also eliminates the dependency to EL module from Kafka nar, because it doesn't have to use EL implementation classes at processor codes.

Tested with standalone and cluster. Confirmed setting properties with EL using variable registry.

I hope this can be merged. Please let me know how you think. Thanks!

JPercivall commented:
When attempting to view state for ConsumeKafka when Kafka is not running, it does not handle it well. In the UI I see "An unexpected error has occurred
An error occurred communicating with the application core. Please check the logs and fix any configuration issues before restarting."

In the logs I see the huge stack trace below that includes viewing and clearing state (I never was able to click clear state though):

2016-08-01 15:21:09,609 WARN [Replicate Request Thread-5] o.a.n.c.c.h.r.ThreadPoolRequestReplicator Failed to replicate request GET /nifi-api/processors/46fc5b3c-0156-1000-8e3a-74a1f65bd5d6/state to localhost:8481 due to {}
com.sun.jersey.api.client.ClientHandlerException: java.net.SocketTimeoutException: Read timed out
	at com.sun.jersey.client.urlconnection.URLConnectionClientHandler.handle(URLConnectionClientHandler.java:155) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.Client.handle(Client.java:652) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.filter.GZIPContentEncodingFilter.handle(GZIPContentEncodingFilter.java:123) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.WebResource.handle(WebResource.java:682) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.WebResource.access$200(WebResource.java:74) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.WebResource$Builder.get(WebResource.java:509) ~[jersey-client-1.19.jar:1.19]
	at org.apache.nifi.cluster.coordination.http.replication.ThreadPoolRequestReplicator.replicateRequest(ThreadPoolRequestReplicator.java:494) ~[nifi-framework-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.cluster.coordination.http.replication.ThreadPoolRequestReplicator$NodeHttpRequest.run(ThreadPoolRequestReplicator.java:688) ~[nifi-framework-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_74]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_74]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_74]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_74]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_74]
Caused by: java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method) ~[na:1.8.0_74]
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116) ~[na:1.8.0_74]
	at java.net.SocketInputStream.read(SocketInputStream.java:170) ~[na:1.8.0_74]
	at java.net.SocketInputStream.read(SocketInputStream.java:141) ~[na:1.8.0_74]
	at sun.security.ssl.InputRecord.readFully(InputRecord.java:465) ~[na:1.8.0_74]
	at sun.security.ssl.InputRecord.read(InputRecord.java:503) ~[na:1.8.0_74]
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:973) ~[na:1.8.0_74]
	at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:930) ~[na:1.8.0_74]
	at sun.security.ssl.AppInputStream.read(AppInputStream.java:105) ~[na:1.8.0_74]
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246) ~[na:1.8.0_74]
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:286) ~[na:1.8.0_74]
	at java.io.BufferedInputStream.read(BufferedInputStream.java:345) ~[na:1.8.0_74]
	at sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:704) ~[na:1.8.0_74]
	at sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:647) ~[na:1.8.0_74]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1536) ~[na:1.8.0_74]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1441) ~[na:1.8.0_74]
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480) ~[na:1.8.0_74]
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:338) ~[na:1.8.0_74]
	at com.sun.jersey.client.urlconnection.URLConnectionClientHandler._invoke(URLConnectionClientHandler.java:253) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.client.urlconnection.URLConnectionClientHandler.handle(URLConnectionClientHandler.java:153) ~[jersey-client-1.19.jar:1.19]
	... 12 common frames omitted
2016-08-01 15:21:09,610 WARN [Replicate Request Thread-5] o.a.n.c.c.h.r.ThreadPoolRequestReplicator 
com.sun.jersey.api.client.ClientHandlerException: java.net.SocketTimeoutException: Read timed out
	at com.sun.jersey.client.urlconnection.URLConnectionClientHandler.handle(URLConnectionClientHandler.java:155) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.Client.handle(Client.java:652) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.filter.GZIPContentEncodingFilter.handle(GZIPContentEncodingFilter.java:123) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.WebResource.handle(WebResource.java:682) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.WebResource.access$200(WebResource.java:74) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.api.client.WebResource$Builder.get(WebResource.java:509) ~[jersey-client-1.19.jar:1.19]
	at org.apache.nifi.cluster.coordination.http.replication.ThreadPoolRequestReplicator.replicateRequest(ThreadPoolRequestReplicator.java:494) ~[nifi-framework-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.cluster.coordination.http.replication.ThreadPoolRequestReplicator$NodeHttpRequest.run(ThreadPoolRequestReplicator.java:688) ~[nifi-framework-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_74]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_74]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_74]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_74]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_74]
Caused by: java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method) ~[na:1.8.0_74]
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116) ~[na:1.8.0_74]
	at java.net.SocketInputStream.read(SocketInputStream.java:170) ~[na:1.8.0_74]
	at java.net.SocketInputStream.read(SocketInputStream.java:141) ~[na:1.8.0_74]
	at sun.security.ssl.InputRecord.readFully(InputRecord.java:465) ~[na:1.8.0_74]
	at sun.security.ssl.InputRecord.read(InputRecord.java:503) ~[na:1.8.0_74]
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:973) ~[na:1.8.0_74]
	at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:930) ~[na:1.8.0_74]
	at sun.security.ssl.AppInputStream.read(AppInputStream.java:105) ~[na:1.8.0_74]
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246) ~[na:1.8.0_74]
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:286) ~[na:1.8.0_74]
	at java.io.BufferedInputStream.read(BufferedInputStream.java:345) ~[na:1.8.0_74]
	at sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:704) ~[na:1.8.0_74]
	at sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:647) ~[na:1.8.0_74]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1536) ~[na:1.8.0_74]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1441) ~[na:1.8.0_74]
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480) ~[na:1.8.0_74]
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:338) ~[na:1.8.0_74]
	at com.sun.jersey.client.urlconnection.URLConnectionClientHandler._invoke(URLConnectionClientHandler.java:253) ~[jersey-client-1.19.jar:1.19]
	at com.sun.jersey.client.urlconnection.URLConnectionClientHandler.handle(URLConnectionClientHandler.java:153) ~[jersey-client-1.19.jar:1.19]
	... 12 common frames omitted
2016-08-01 15:21:11,194 INFO [Process Cluster Protocol Request-1] o.a.n.c.p.impl.SocketProtocolListener Finished processing request 77227f6d-5ba1-4b44-b8c1-d084c85a6f77 (type=HEARTBEAT, length=1320 bytes) from localhost in 20 millis
2016-08-01 15:21:12,305 INFO [Heartbeat Monitor Thread-1] o.a.n.c.c.h.AbstractHeartbeatMonitor Finished processing 3 heartbeats in 18563 nanos
2016-08-01 15:21:12,392 INFO [Clustering Tasks Thread-1] org.apache.nifi.cluster.heartbeat Heartbeat created at 2016-08-01 15:21:12,371 and sent to localhost:6001 at 2016-08-01 15:21:12,392; send took 20 millis
2016-08-01 15:21:12,393 INFO [Process Cluster Protocol Request-4] o.a.n.c.p.impl.SocketProtocolListener Finished processing request a921d17d-df8e-4a55-a2a4-17358f6325e6 (type=HEARTBEAT, length=1381 bytes) from localhost in 20 millis
2016-08-01 15:21:12,733 INFO [Process Cluster Protocol Request-3] o.a.n.c.p.impl.SocketProtocolListener Finished processing request ecaa553d-7ab6-4e1c-9087-6fc17bd746c9 (type=HEARTBEAT, length=1320 bytes) from localhost in 29 millis
2016-08-01 15:21:13,799 INFO [Provenance Maintenance Thread-3] o.a.n.p.PersistentProvenanceRepository Created new Provenance Event Writers for events starting with ID 186
2016-08-01 15:21:13,842 INFO [Provenance Repository Rollover Thread-1] o.a.n.p.PersistentProvenanceRepository Successfully merged 16 journal files (68 records) into single Provenance Log File ./provenance_repository/118.prov in 45 milliseconds
2016-08-01 15:21:13,842 INFO [Provenance Repository Rollover Thread-1] o.a.n.p.PersistentProvenanceRepository Successfully Rolled over Provenance Event file containing 161 records
2016-08-01 15:21:14,610 ERROR [NiFi Web Server-23] o.a.nifi.web.StandardNiFiServiceFacade Failed to clear state for ConsumeKafka[id=46fc5b3c-0156-1000-8e3a-74a1f65bd5d6] due to java.io.IOException: Fetch offsets failed due to java.util.concurrent.TimeoutException
java.io.IOException: Fetch offsets failed due to java.util.concurrent.TimeoutException
	at org.apache.nifi.processors.kafka.pubsub.ConsumeKafka.submitConsumerGroupCommand(ConsumeKafka.java:486) ~[na:na]
	at org.apache.nifi.processors.kafka.pubsub.ConsumeKafka.getExternalState(ConsumeKafka.java:353) ~[na:na]
	at org.apache.nifi.web.dao.impl.StandardComponentStateDAO.getExternalState(StandardComponentStateDAO.java:59) ~[classes/:na]
	at org.apache.nifi.web.dao.impl.StandardComponentStateDAO.getState(StandardComponentStateDAO.java:38) ~[classes/:na]
	at org.apache.nifi.web.StandardNiFiServiceFacade.getComponentState(StandardNiFiServiceFacade.java:1983) [classes/:1.0.0-SNAPSHOT]
	at org.apache.nifi.web.StandardNiFiServiceFacade.getProcessorState(StandardNiFiServiceFacade.java:2000) [classes/:1.0.0-SNAPSHOT]
	at org.apache.nifi.web.StandardNiFiServiceFacade$$FastClassBySpringCGLIB$$358780e0.invoke(<generated>) [classes/:1.0.0-SNAPSHOT]
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204) [spring-core-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:720) [spring-aop-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157) [spring-aop-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:85) [spring-aop-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.apache.nifi.web.NiFiServiceFacadeLock.getLock(NiFiServiceFacadeLock.java:162) [classes/:1.0.0-SNAPSHOT]
	at sun.reflect.GeneratedMethodAccessor77.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_74]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_74]
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:621) [spring-aop-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:610) [spring-aop-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:68) [spring-aop-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) [spring-aop-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92) [spring-aop-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) [spring-aop-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:655) [spring-aop-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.apache.nifi.web.StandardNiFiServiceFacade$$EnhancerBySpringCGLIB$$bb517908.getProcessorState(<generated>) [classes/:1.0.0-SNAPSHOT]
	at org.apache.nifi.web.api.ProcessorResource.getState(ProcessorResource.java:312) [classes/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_74]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_74]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_74]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_74]
	at com.sun.jersey.spi.container.JavaMethodInvokerFactory$1.invoke(JavaMethodInvokerFactory.java:60) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.model.method.dispatch.AbstractResourceMethodDispatchProvider$ResponseOutInvoker._dispatch(AbstractResourceMethodDispatchProvider.java:205) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.model.method.dispatch.ResourceJavaMethodDispatcher.dispatch(ResourceJavaMethodDispatcher.java:75) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.HttpMethodRule.accept(HttpMethodRule.java:302) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.ResourceClassRule.accept(ResourceClassRule.java:108) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.RootResourceClassesRule.accept(RootResourceClassesRule.java:84) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1542) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1473) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409) [jersey-servlet-1.19.jar:1.19]
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:558) [jersey-servlet-1.19.jar:1.19]
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:733) [jersey-servlet-1.19.jar:1.19]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) [javax.servlet-api-3.1.0.jar:3.1.0]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:845) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1689) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlets.GzipFilter.doFilter(GzipFilter.java:51) [jetty-servlets-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.apache.nifi.web.filter.RequestLogger.doFilter(RequestLogger.java:66) [classes/:na]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:316) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:126) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:90) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:122) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.apache.nifi.web.security.NiFiAuthenticationFilter.doFilter(NiFiAuthenticationFilter.java:59) [nifi-web-security-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.apache.nifi.web.security.NiFiAuthenticationFilter.doFilter(NiFiAuthenticationFilter.java:59) [nifi-web-security-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.apache.nifi.web.security.NiFiAuthenticationFilter.authenticate(NiFiAuthenticationFilter.java:83) [nifi-web-security-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.web.security.NiFiAuthenticationFilter.doFilter(NiFiAuthenticationFilter.java:57) [nifi-web-security-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:213) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:176) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:346) [spring-web-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:262) [spring-web-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.apache.nifi.web.filter.TimerFilter.doFilter(TimerFilter.java:51) [classes/:na]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:581) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548) [jetty-security-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1174) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:511) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1106) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:119) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.Server.handle(Server.java:524) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:319) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:253) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273) [jetty-io-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95) [jetty-io-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:186) [jetty-io-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273) [jetty-io-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95) [jetty-io-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93) [jetty-io-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_74]
Caused by: java.util.concurrent.TimeoutException: null
	at java.util.concurrent.FutureTask.get(FutureTask.java:205) ~[na:1.8.0_74]
	at org.apache.nifi.processors.kafka.pubsub.ConsumeKafka.submitConsumerGroupCommand(ConsumeKafka.java:476) ~[na:na]
	... 97 common frames omitted

JPercivall commented:
Also would it make sense to have in the "View State" menu, instead of just "External" for clustered external state make it "External - Clustered"?

JPercivall commented:
@olegz @ijokarumawak I don't think "Queue Buffering Max Time" is working properly. I was getting super poor performance out of PutKafka and couldn't understand why it would take ~5 seconds for a single message to be sent. I then found this property and when I adjusted it lower it fixed my performance problems (and higher according made it worse).

Looking at the property descriptor and corresponding property in Kafka ([seach for "linger.ms"](http://kafka.apache.org/documentation.html)), I would expect it to batch multiple flowfiles but instead it just hung and only processed one flowfile per batch. Below are a couple screenshots of the configuration and the stats of running 3 node cluster (saw the saw phenomenon running primary node only and the scheduling period was "0 secs"):

![screen shot 2016-08-01 at 5 22 23 pm](https://cloud.githubusercontent.com/assets/11302527/17309375/8d1189e0-580c-11e6-9c58-cf1a8f9bbc99.png)

![screen shot 2016-08-01 at 5 12 26 pm](https://cloud.githubusercontent.com/assets/11302527/17309350/792e9512-580c-11e6-8d70-19233f7132e7.png)

JPercivall commented:
Also I'm not sure if this is due to Kafka or our implementation but it seems the offsets for GetKafka don't get updated until the processor is stopped. I will start it, after a while it will have processed some data but when I view the state I see nothing. As soon as I stop it and view the state I see the offsets. In the ZK output I see these 4 statements each time state is requested:

[2016-08-01 17:31:50,591] INFO Accepted socket connection from /127.0.0.1:59809 (org.apache.zookeeper.server.NIOServerCnxnFactory)
[2016-08-01 17:31:50,591] INFO Client attempting to establish new session at /127.0.0.1:59809 (org.apache.zookeeper.server.ZooKeeperServer)
[2016-08-01 17:31:50,593] INFO Established session 0x15647e175570031 with negotiated timeout 30000 for client /127.0.0.1:59809 (org.apache.zookeeper.server.ZooKeeperServer)
[2016-08-01 17:31:50,597] INFO Processed session termination for sessionid: 0x15647e175570031 (org.apache.zookeeper.server.PrepRequestProcessor)
[2016-08-01 17:31:50,598] INFO Closed socket connection for client /127.0.0.1:59809 which had sessionid 0x15647e175570031 (org.apache.zookeeper.server.NIOServerCnxn)

JPercivall commented:
Continuing the comment on "linger.ms", it appears when the "linger.ms" property is used in PublishKafka it yields the same results. Going forward it looks like we should be taking two steps:

1: Change our default in PutKafka to match the Kafka default. Looking at the [Kafka documentation](http://kafka.apache.org/documentation.html) the default value for "linger.ms" is 0 seconds, for no delay.

2: Figure out whether the mishandling of the property is a problem in our Processors or the underlying libraries. This may be tricky to do and I can start but I will probably need help. @olegz could you help here?

JPercivall commented:
@ijokarumawak there is a lot of great work here but the various tickets that are included in this one PR are making it difficult to review and test. Could we close this PR and break it out into separate ones for each ticket so that they can be addressed individually? That would make it much more efficient to get the bugs fixes into master and properly test the new external state management.

Also could we move NIFI-2078 out of 1.0 to 1.1? This is an important new feature with API implications and it would be nice to finishing reviewing once the other API changes take effect and the various UI blocking bugs are addressed.

ijokarumawak commented:
@JPercivall Thanks for taking your time to review it, really appreciated. I agree with breaking this into separate PRs, and moving it to 1.1. I think only NIFI-2441: View state JS issue. I will close this PR and going to submit separate PRs.
